<div style="text-align: center; padding: 30px; background-color: #f5f7fa; border-radius: 12px; box-shadow: 0px 4px 10px rgba(0,0,0,0.1);">

  <h1 style="color: #2c3e50; font-family: 'Helvetica Neue', sans-serif; font-size: 42px; margin-bottom: 10px;">
    Virtual Lab 10/11 Project
  </h1>

  <h3 style="color: #7f8c8d; font-family: 'Georgia', serif; font-weight: normal; font-size: 22px; margin-top: 0;">
    Network and Pathway-Based Functional Prediction Tasks
  </h3>

  <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 16px; margin-top: 20px;">
    Author: <b>Vallari Eastman</b> <br>
    Due Date: <b>Monday December 1st, 2025 11:59 p.m. ET</b>
  </p>
</div>

<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    Introduction:
  </h2>
<p style="text-indent:30px; color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px; line-height:1.6;">
    Post-translational modifications (PTMs) play a central role in regulating neuronal development, signaling, and structural remodeling in the brain. Large-scale proteomics datasets allow these modifications to be mapped at peptide-level resolution, enabling the identification of regulated proteins and pathways across developmental stages. This lab involves integrating peptide sequences, PTM annotations, and protein-level information from two complementary datasets to characterize PTM patterns and link them to biological function, signaling pathways, and network context.
</p>

<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    Resources Needed to Complete the Lab:
  </h2>
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> CompOmics Tutorial</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       Aaccess the tutorial through Componics website using the following link: https://www.compomics.com/bioinformatics-for-proteomics<br><br>
1. Identify peptides, proteins, and their modifications<br>
2. Annotate the data with existing biological knowledge<br>
3. Share the data using online repositories<br>
<br>
   </p>
   <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b>Publication:</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       Edwards, A. V., Schwämmle, V., & Larsen, M. R. (2014). Neuronal process structure and growth proteins are targets of heavy PTM regulation during brain development. Journal of Proteomics, 101, 77-87.   JHU Library PermalinkLinks to an external site.<br>
In this study, the researchers use mass spectrometry and proteomic strategies to present biological data showing spatiotemporal PTM co-localization across multiple PTM categories, which display changes over development of the brain.<br><br>
   </p>
   <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b>Associated Data Sets</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       The example files below come from the following PRIDE repository:<br>
https://www.ebi.ac.uk/pride/archive/projects/PXD003932<br><br>
       <b>Dataset 1: Peptide Sequences.xlsx</b><br>
       All confidently assigned peptide sequences and PTMs detected in mouse brain PTM mapping data. Data has been processed to remove redundancy and accumulate iTRAQ ratios using ReportSites (Edwards et al, 2012). Annotation - de = deamidation (putative glycosylation), ph = phosphorylation, ac = acetylation.<br><br>
       <b>Dataset 2: Regulated Peptides.xlsx</b><br>
       Compiled regulated peptides from mouse brain developmental data.  'Cluster' refers to peptides displaying similar abundance profiles as defined by the clustering algorithm used; Max Membership is a measure of the 'fit' of this peptide to the cluster; regulation at Day 8, 21, or 80 versus Day 0 is displayed as 'True' or 'False' in columns E to G; individual replicate peptide abundance information is given in columns H through W in log2 format; log2 values of the means of these abundance values are given in columns  X to AA; protein ID's are given as uniprot accessions in column AB; pRS probabilities are given in column AC for those peptides for which it is relevant; protein names are given in column AD; the total number of PTMs on the peptide in questoin is given in columns AE to AG.
   </p>
</div>

<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    1. Identify peptides, proteins, and their modifications
  </h2>
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step A. Identify peptides</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       <b>Use Dataset 1 (Peptide Sequences.xlsx):</b><br>
       1. Extract all peptide sequences<br>
       2. Remove duplicates (they represent redundant identifications removed by ReportSites)<br>
       3. Count total peptides<br>
       4. Count per PTM category<br>
   </p>
</div>

In [2]:
# Load the dataset

import pandas as pd

df1 = pd.read_excel("Peptide Sequences.xlsx", skiprows=5)

# Clean up column names: remove leading/trailing spaces and line breaks
#df1.columns = df.columns.str.strip()
df1.columns = df1.columns.str.strip().str.replace('\n', ' ', regex=True)

# drop all "Unnamed" columns
df1 = df1.loc[:, ~df1.columns.str.contains("Unnamed")]

# Inspect columns to see what information the file has
print("Columns:", df1.columns.tolist())
df1.head()

Columns: ['Protein Id', 'Starting Residue', 'Ending Residue', 'Peptide Sequence', 'Mean', 'Standard Deviation', 'Minimum', 'Maximum', 'Mean.1', 'Standard Deviation.1', 'Minimum.1', 'Maximum.1', 'Mean.2', 'Standard Deviation.2', 'Minimum.2', 'Maximum.2']


,Protein Id,Starting Residue,Ending Residue,Peptide Sequence,Mean,Standard Deviation,Minimum,Maximum,Mean.1,Standard Deviation.1,Minimum.1,Maximum.1,Mean.2,Standard Deviation.2,Minimum.2,Maximum.2
0,P43006,209,226,AVISMLN(de)ETMNEAPEETK,1.059,1.473,-2.191,6.325,2.109,2.036,-3.373,6.745,2.565,2.256,-3.71,6.48
1,P14094,255,274,YLQPLLAVQFTN(de)LTVDTEIR,0.345,1.101,-2.962,2.948,0.666,1.772,-4.175,6.104,0.662,1.96,-4.968,6.198
2,P04370,229,236,T(ph)PPPSQGK,0.814,1.446,-3.563,5.037,1.887,2.199,-3.971,7.721,2.433,2.326,-4.041,8.257
3,P54227,30,41,ESVPDFPLS(ph)PPK,-0.679,1.727,-9.498,3.988,-1.322,1.861,-7.601,4.349,-1.524,1.92,-6.77,4.268
4,Q9JIS5,492,507,VEHVTFN(de)FTLENQIHR,1.386,1.23,-1.419,5.552,1.446,1.457,-1.745,6.745,1.561,1.699,-2.771,6.48


In [3]:
# Extract peptide sequences

# Replace column name if your file uses something slightly different
peptide_col = "Peptide Sequence"

peptides = df1[peptide_col]
print("Total peptide entries (including duplicates):", peptides.shape[0])

Total peptide entries (including duplicates): 16400


In [4]:
#Remove duplicates & Count total peptide sequences

# Step 1: Extract the peptide sequence column
peptides = df1["Peptide Sequence"]

# Step 2: Remove duplicates
unique_peptides = peptides.drop_duplicates().reset_index(drop=True)

# Step 3: Print count
print("Unique peptide sequences:", unique_peptides.shape[0])

Unique peptide sequences: 15657


In [5]:
# Count peptides per PTM category (PTMs are inside the peptide sequence)

import re

total_peptides = df1.shape[0]

def has_ph(seq):
    return "ph" in seq.lower()

def has_ac(seq):
    return "ac" in seq.lower()

def has_de(seq):
    return "de" in seq.lower()

df1['has_ph'] = df1['Peptide Sequence'].apply(has_ph)
df1['has_ac'] = df1['Peptide Sequence'].apply(has_ac)
df1['has_de'] = df1['Peptide Sequence'].apply(has_de)

summary = pd.DataFrame({
    "PTM Type": ["Phosphorylated", "Acetylated", "Deamidated"],
    "Count": [df1['has_ph'].sum(), df1['has_ac'].sum(), df1['has_de'].sum()]
})

summary["Percent (%)"] = (summary["Count"] / total_peptides * 100).round(2)

# Add total row
total_row = pd.DataFrame({
    "PTM Type": ["Total peptides"],
    "Count": [total_peptides],
    "Percent (%)": [100.0]
})

summary_with_total = pd.concat([summary, total_row], ignore_index=True)
summary_with_total

,PTM Type,Count,Percent (%)
0,Phosphorylated,12969,79.08
1,Acetylated,1757,10.71
2,Deamidated,3347,20.41
3,Total peptides,16400,100.00


<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step B. Identify proteins</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       <b>Use Dataset 2 (Regulated Peptides.xls):</b><br>
       1. Extract all UniProt IDs (AB column)<br>
       2. Remove duplicates<br>
       3. This becomes a list of “identified proteins”<br>
       4. Merge Dataset 1 + Dataset 2 by peptide sequence to link peptides → proteins consistently.<br>
   </p>
</div>

In [6]:
# Load Dataset & Clean Columns

# Load Dataset 2
df2 = pd.read_excel("Regulated Peptides.xlsx", skiprows=7)

# Clean column names: strip spaces and line breaks
# df2.columns = df2.columns.str.strip()
df2.columns = df2.columns.str.strip().str.replace('\n', ' ', regex=True)

# drop all "Unnamed" columns
df2 = df2.loc[:, ~df2.columns.str.contains("Unnamed")]

print("Columns:", df2.columns.tolist())
df2.head()


Columns: ['Peptide sequence', 'Cluster', 'Max Membership', 'D8 vs. D0', 'D21 vs. D0', 'D80 vs. D0', '114 Rep 1', '115 Rep 1', '116 Rep 1', '117 Rep 1', '114 Rep 2', '115 Rep 2', '116 Rep 2', '117 Rep 2', '114 Rep 3', '115 Rep 3', '116 Rep 3', '117 Rep 3', '114 Rep 4', '115 Rep 4', '116 Rep 4', '117 Rep 4', 'Mean of log2 D0', 'Mean of log2 D8', 'Mean of log2 Day 21', 'Mean of log2 Day 80', 'Protein ID', 'pRS Site Probabilities', 'Protein Name', 'Phospho', 'Acetyl', 'Glyco (put.)']


,Peptide sequence,Cluster,Max Membership,D8 vs. D0,D21 vs. D0,D80 vs. D0,114 Rep 1,115 Rep 1,116 Rep 1,117 Rep 1,...,Mean of log2 D0,Mean of log2 D8,Mean of log2 Day 21,Mean of log2 Day 80,Protein ID,pRS Site Probabilities,Protein Name,Phospho,Acetyl,Glyco (put.)
0,QPLLLSphEDEEDTK,5,0.2496,True,False,False,NaN,NaN,NaN,NaN,...,-2.31654,2.24266,-0.61161,0.51602,B5DFC8,S(6): 100.0; T(12): 0.0,Eukaryotic translation initiation factor 3 sub...,1,0,0
1,EGDGSATTDAAPATSphPKAEEPSK,6,0.61381,True,False,True,3.19972,-1.7694,0.44518,-1.34868,...,1.60449,-0.91106,-0.08477,-0.46133,P06837,S(5): 0.0; T(7): 0.0; T(8): 0.0; T(14): 0.6; S...,Neuromodulin,1,0,0
2,SKIGSTphENLK,5,0.26915,True,False,True,-1.40737,1.12034,-1.79301,0.87568,...,-1.82987,1.64793,-1.25073,0.78517,P10637;Q6TS35,S(1): 0.0; S(5): 99.5; T(6): 0.5,Microtubule-associated protein tau,1,0,0
3,KLGGDVSphPTQIDVSQFGSFK,6,0.64358,True,True,True,3.634,-2.21603,0.15517,-1.25565,...,1.52634,-0.7291,-0.20696,-0.54614,P14873,S(7): 98.7; T(9): 1.3; S(14): 0.0; S(18): 0.0,MAP-1B,1,0,0
4,KLGGDVSPTphQIDVSQFGSFK,6,0.63335,True,True,True,3.49034,-2.32651,0.34184,-1.18818,...,2.10643,-1.02612,-0.35179,-0.68438,P14873,S(7): 0.6; T(9): 99.4; S(14): 0.0; S(18): 0.0,MAP-1B,1,0,0


In [7]:
# Extract all UniProt IDs (AB column) and remove duplicates

# Try to detect a UniProt / Protein ID column automatically
candidate_cols = [c for c in df2.columns if "uniprot" in c.lower() or "protein id" in c.lower()]

print("Candidate protein ID columns:", candidate_cols)

Candidate protein ID columns: ['Protein ID']


In [8]:
uniprot_col = candidate_cols[0]  # e.g. "Protein ID"

uniprot_ids = df2[uniprot_col]

print("Total IDs (with duplicates):", uniprot_ids.shape[0])

Total IDs (with duplicates): 480


In [9]:
# Remove Duplicates
unique_uniprot_ids = uniprot_ids.dropna().drop_duplicates().reset_index(drop=True)

print("Unique UniProt IDs:", unique_uniprot_ids.shape[0])
print("Total IDs (with duplicates & NAs):", uniprot_ids.shape[0])

# small DataFrame:
identified_proteins = pd.DataFrame({"UniProt ID": unique_uniprot_ids})
identified_proteins.head()

Unique UniProt IDs: 186
Total IDs (with duplicates & NAs): 480


,UniProt ID
0,B5DFC8
1,P06837
2,P10637;Q6TS35
3,P14873
4,P14873;P15205


In [10]:
import re
import pandas as pd

def canonical_peptide(seq):
    """
    Turn any peptide string with PTM annotations into plain AA sequence.
    Keeps only A–Z, throws away ph/ac/de, parentheses, etc.
    """
    if not isinstance(seq, str):
        return None
    seq = seq.upper()
    # remove explicit ptm tags like PH, AC, DE in df2 *after* uppercasing
    seq = re.sub(r"(PH|AC|DE)", "", seq)
    # now keep only A-Z
    return re.sub(r"[^A-Z]", "", seq)

# For Dataset 1
df1["pep_clean"] = df1["Peptide Sequence"].apply(canonical_peptide)

# For Dataset 2
df2["pep_clean"] = df2["Peptide sequence"].apply(canonical_peptide)

df1[["Peptide Sequence", "pep_clean"]].head()
df2[["Peptide sequence", "pep_clean"]].head()

,Peptide sequence,pep_clean
0,QPLLLSphEDEEDTK,QPLLLSEEDTK
1,EGDGSATTDAAPATSphPKAEEPSK,EGDGSATTDAAPATSPKAEEPSK
2,SKIGSTphENLK,SKIGSTENLK
3,KLGGDVSphPTQIDVSQFGSFK,KLGGDVSPTQIDVSQFGSFK
4,KLGGDVSPTphQIDVSQFGSFK,KLGGDVSPTQIDVSQFGSFK


In [11]:
set1 = set(df1["pep_clean"].dropna().unique())
set2 = set(df2["pep_clean"].dropna().unique())

len(set1 & set2), len(set1), len(set2)

(326, 13012, 329)

In [12]:
pep_protein_map = (
    df2[["pep_clean", "Protein ID"]]
    .dropna(subset=["Protein ID"])
    .drop_duplicates(subset=["pep_clean"])
)

pep_protein_map.head()

,pep_clean,Protein ID
0,QPLLLSEEDTK,B5DFC8
1,EGDGSATTDAAPATSPKAEEPSK,P06837
2,SKIGSTENLK,P10637;Q6TS35
3,KLGGDVSPTQIDVSQFGSFK,P14873
5,DYNASASTISPPSSMEEDK,P14873;P15205


In [13]:
# Merge Dataset 1 + Dataset 2 by peptide sequence

df1_annotated = df1.merge(
    pep_protein_map,
    on="pep_clean",   # same column name in both now
    how="left"
)

df1_annotated.head()

,Protein Id,Starting Residue,Ending Residue,Peptide Sequence,Mean,Standard Deviation,Minimum,Maximum,Mean.1,Standard Deviation.1,...,Maximum.1,Mean.2,Standard Deviation.2,Minimum.2,Maximum.2,has_ph,has_ac,has_de,pep_clean,Protein ID
0,P43006,209,226,AVISMLN(de)ETMNEAPEETK,1.059,1.473,-2.191,6.325,2.109,2.036,...,6.745,2.565,2.256,-3.71,6.48,False,False,True,AVISMLNETMNEAPEETK,NaN
1,P14094,255,274,YLQPLLAVQFTN(de)LTVDTEIR,0.345,1.101,-2.962,2.948,0.666,1.772,...,6.104,0.662,1.96,-4.968,6.198,False,False,True,YLQPLLAVQFTNLTVDTEIR,NaN
2,P04370,229,236,T(ph)PPPSQGK,0.814,1.446,-3.563,5.037,1.887,2.199,...,7.721,2.433,2.326,-4.041,8.257,True,False,False,TPPPSQGK,NaN
3,P54227,30,41,ESVPDFPLS(ph)PPK,-0.679,1.727,-9.498,3.988,-1.322,1.861,...,4.349,-1.524,1.92,-6.77,4.268,True,False,False,ESVPDFPLSPPK,P54227
4,Q9JIS5,492,507,VEHVTFN(de)FTLENQIHR,1.386,1.23,-1.419,5.552,1.446,1.457,...,6.745,1.561,1.699,-2.771,6.48,False,False,True,VEHVTFNFTLENQIHR,NaN


In [14]:
mapped = df1_annotated["Protein ID"].notna().sum()
total  = df1_annotated.shape[0]

print(f"Peptides with mapped UniProt ID: {mapped} / {total}")

Peptides with mapped UniProt ID: 667 / 16400


In [15]:
len(set(df1["pep_clean"]))  # unique sequences in df1
len(set(df2["pep_clean"]))  # unique sequences in df2
len(set(df1["pep_clean"]) & set(df2["pep_clean"]))  # unique canonical matches

326

In [16]:
df2_matches = df2[df2["pep_clean"].isin(df1["pep_clean"])]
df2_matches.shape

(471, 33)

<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step C. Identify Modifications (PTMs)</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       <b>Dataset 1 contains the raw PTM annotations on each peptide:</b><br>
PTM codes:<br>
ph = phosphorylation<br>
ac = acetylation<br>
de = deamidation (putative glycosylation)<br>
<br>
<b>Dataset 2 has PTM counts per peptide (columns AE–AG)</b><br>
<br>
1. Count PTMs by type: phosphorylation / acetylation / deamidation<br>
2. Determine how many peptides contain multiple PTMs<br>
3. Compare regulated vs non-regulated PTMs (Dataset 2 = regulated only)<br>
   </p>
</div>

In [17]:
# Count PTMs by Type

# How many peptides in df1 have at least one of each PTM type?
ptm_counts_df1 = pd.DataFrame({
    "PTM type": ["Phosphorylated", "Acetylated", "Deamidated"],
    "Peptides with ≥1 site": [
        df1["has_ph"].sum(),
        df1["has_ac"].sum(),
        df1["has_de"].sum()
    ]
})

ptm_counts_df1

total_peptides_df1 = df1.shape[0]
ptm_counts_df1["Percent of all peptides (%)"] = (
    ptm_counts_df1["Peptides with ≥1 site"] / total_peptides_df1 * 100
).round(2)

ptm_counts_df1


,PTM type,Peptides with ≥1 site,Percent of all peptides (%)
0,Phosphorylated,12969,79.08
1,Acetylated,1757,10.71
2,Deamidated,3347,20.41


In [18]:
# Clean column names
df2.columns = df2.columns.str.strip()

# PTM columns
ptm_cols = ["Phospho", "Acetyl", "Glyco (put.)"]

# Convert PTM columns to numeric
df2[ptm_cols] = df2[ptm_cols].apply(pd.to_numeric, errors='coerce')

# Now compute site counts
ptm_site_counts_df2 = df2[ptm_cols].sum()
ptm_peptide_counts_df2 = (df2[ptm_cols] > 0).sum()

ptm_site_counts_df2, ptm_peptide_counts_df2

(Phospho         393.0
 Acetyl          100.0
 Glyco (put.)     30.0
 dtype: float64,
 Phospho         356
 Acetyl           97
 Glyco (put.)     27
 dtype: int64)

In [19]:
# Determine how many have multiple PTMs

# Number of PTM types on each peptide (0–3)
df1["n_ptm_types"] = df1[["has_ph", "has_ac", "has_de"]].sum(axis=1)

n_multi_df1 = (df1["n_ptm_types"] >= 2).sum()
n_any_ptm_df1 = (df1["n_ptm_types"] >= 1).sum()

print("Peptides with ≥1 PTM type:", n_any_ptm_df1)
print("Peptides with ≥2 PTM types:", n_multi_df1)

multi_summary_df1 = df1["n_ptm_types"].value_counts().sort_index()
multi_summary_df1

# True if peptide has at least one site of each type
df2["has_ph"] = df2["Phospho"] > 0
df2["has_ac"] = df2["Acetyl"] > 0
df2["has_de"] = df2["Glyco (put.)"] > 0

# Number of PTM types per regulated peptide
df2["n_ptm_types"] = df2[["has_ph", "has_ac", "has_de"]].sum(axis=1)

n_multi_df2 = (df2["n_ptm_types"] >= 2).sum()
n_any_ptm_df2 = (df2["n_ptm_types"] >= 1).sum()

print("Regulated peptides with ≥1 PTM type:", n_any_ptm_df2)
print("Regulated peptides with ≥2 PTM types:", n_multi_df2)

df2["n_ptm_types"].value_counts().sort_index()


Peptides with ≥1 PTM type: 16400
Peptides with ≥2 PTM types: 1635
Regulated peptides with ≥1 PTM type: 474
Regulated peptides with ≥2 PTM types: 6


n_ptm_types
0      6
1    468
2      6
Name: count, dtype: int64

In [20]:
# Compare regulated vs non-regulated PTMs

# Set of regulated canonical sequences
regulated_set = set(df2["pep_clean"].dropna())

# Boolean flag on df1
df1["is_regulated"] = df1["pep_clean"].isin(regulated_set)

df1["is_regulated"].value_counts()

def summarize_ptms(group):
    return pd.Series({
        "N peptides": group.shape[0],
        "N ph peptides": group["has_ph"].sum(),
        "N ac peptides": group["has_ac"].sum(),
        "N de peptides": group["has_de"].sum()
    })

#ptm_reg_summary = df1.groupby("is_regulated").apply(summarize_ptms)
ptm_reg_summary = df1.groupby("is_regulated", group_keys=True).apply(
    summarize_ptms,
    include_groups=False
)

# Make the index human-readable
ptm_reg_summary.index = ["Non-regulated", "Regulated"]

# Add percentages within each group
for col in ["N ph peptides", "N ac peptides", "N de peptides"]:
    ptm_reg_summary[col + " (%)"] = (
        ptm_reg_summary[col] / ptm_reg_summary["N peptides"] * 100
    ).round(2)

ptm_reg_summary

,N peptides,N ph peptides,N ac peptides,N de peptides,N ph peptides (%),N ac peptides (%),N de peptides (%)
Non-regulated,15733,12492,1594,3238,79.40,10.13,20.58
Regulated,667,477,163,109,71.51,24.44,16.34


<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    2. Annotate the data with existing biological knowledge
  </h2>
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step A. Add biological annotation (UniProt)</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       <b>Helpful Tools:</b><br>
       1. UniProt batch search<br>
       2. DAVID<br>
       3. g:Profiler<br>
       4. Reactome Pathway Browser<br>
       5. STRING interaction network<br><br>
       <b>Annotations:</b><br>
       1. Protein function<br>
       From UniProt: synaptic proteins, cytoskeleton regulators, neuronal signaling proteins, etc.<br><br>
       2. PTM biology<br>
       Phosphorylation → kinases, signaling, growth cone dynamics<br>
       Acetylation → chromatin, metabolic enzymes<br>
       Deamidation/glycosylation → extracellular proteins, secretory pathways<br><br>
       3. Pathways & processes<br>
Axon growth and guidance<br>
Cytoskeletal remodeling<br>
Vesicle transport<br>
Neuronal differentiation<br>
Synaptic plasticity<br><br>
4. Network context<br>
Upload UniProt IDs to STRING:<br>
   </p>
</div>

In [23]:
# Creates a protein-level summary to annotate

# Mapped peptides only
mapped = df1_annotated[df1_annotated["Protein ID"].notna()].copy()

# Aggregate PTM info to the protein level
protein_summary = mapped.groupby("Protein ID").agg(
    N_peptides    = ("pep_clean", "nunique"),
    N_ph_pep      = ("has_ph", "sum"),
    N_ac_pep      = ("has_ac", "sum"),
    N_de_pep      = ("has_de", "sum")
)

# Boolean: does this protein ever appear with each PTM type?
protein_summary["has_ph"] = protein_summary["N_ph_pep"] > 0
protein_summary["has_ac"] = protein_summary["N_ac_pep"] > 0
protein_summary["has_de"] = protein_summary["N_de_pep"] > 0

protein_summary.head()


,N_peptides,N_ph_pep,N_ac_pep,N_de_pep,has_ph,has_ac,has_de
Protein ID,,,,,,,
A1L108,1,1,0,0,True,False,False
A2AJI0,1,2,0,0,True,False,False
A2AN08,1,1,0,0,True,False,False
B5DFC8,1,1,0,1,True,False,True
O08539,1,1,0,0,True,False,False


In [25]:
# Export UniProt IDs for batch annotation

protein_ids = protein_summary.index.to_series().reset_index(drop=True)
protein_ids.to_csv("protein_ids_for_uniprot.txt", index=False, header=False)

# Batch ID Mapping Unitprot

1. Go to: https://www.uniprot.org/id-mapping
2. Use file created earlier and upload list IDs
3. "From:" UniProtKB AC/ID was selected
4. "To:" UniProtKB was selected
5. Name project "uniprot_annotations"
6. Save results as tsv

In [26]:
# Bring Annotations Back into Jupyter

uniprot = pd.read_csv("uniprot_annotations.tsv", sep="\t")

# Make sure the ID column matches Protein ID naming
# Often the UniProt column is called "Entry"
uniprot = uniprot.rename(columns={"Entry": "Protein ID"})

# Merge with your protein summary
protein_annotated = protein_summary.merge(uniprot, on="Protein ID", how="left")

protein_annotated.head()

,Protein ID,N_peptides,N_ph_pep,N_ac_pep,N_de_pep,has_ph,has_ac,has_de,From,Reviewed,Entry Name,Protein names,Gene Names,Organism,Length
0,A1L108,1,1,0,0,True,False,False,A1L108,reviewed,ARP5L_RAT,Actin-related protein 2/3 complex subunit 5-li...,Arpc5l,Rattus norvegicus (Rat),153.0
1,A2AJI0,1,2,0,0,True,False,False,A2AJI0,reviewed,MA7D1_MOUSE,MAP7 domain-containing protein 1,Map7d1 Kiaa1187 Mtap7d1,Mus musculus (Mouse),846.0
2,A2AN08,1,1,0,0,True,False,False,A2AN08,reviewed,UBR4_MOUSE,E3 ubiquitin-protein ligase UBR4 (EC 2.3.2.27)...,Ubr4 Kiaa0462,Mus musculus (Mouse),5180.0
3,B5DFC8,1,1,0,1,True,False,True,B5DFC8,reviewed,EIF3C_RAT,Eukaryotic translation initiation factor 3 sub...,Eif3c Eif3s8,Rattus norvegicus (Rat),911.0
4,O08539,1,1,0,0,True,False,False,O08539,reviewed,BIN1_MOUSE,Myc box-dependent-interacting protein 1 (Amphi...,Bin1 Amphl Sh3p9,Mus musculus (Mouse),588.0


<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step B. Annotate peptides with PTM interpretation</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       <b>Annotations:</b><br>
       1. Whether the PTM is on a known regulatory site<br>
       2. Whether the PTM is conserved across species<br>
       3. Whether multiple PTMs co-occur (crosstalk sites)<br>
       4. Whether the PTM site appears on proteins involved in neural development<br>
   </p>
</div>

In [27]:
# Use earlier PTM flags to add categorical annotations

def ptm_category(row):
    types = []
    if row["has_ph"]:
        types.append("phospho")
    if row["has_ac"]:
        types.append("acetyl")
    if row["has_de"]:
        types.append("deamidation")
    if not types:
        return "no PTM"
    if len(types) == 1:
        return types[0]
    return "multi-PTM (" + "+".join(types) + ")"

protein_annotated["PTM_category"] = protein_annotated.apply(ptm_category, axis=1)
protein_annotated["PTM_category"].value_counts()

PTM_category
phospho                                   86
acetyl                                    38
multi-PTM (phospho+deamidation)           31
deamidation                               11
multi-PTM (phospho+acetyl)                 7
multi-PTM (acetyl+deamidation)             6
multi-PTM (phospho+acetyl+deamidation)     5
Name: count, dtype: int64

In [38]:
# Export list for Reactome (Requires one protein per line)

# Start from your merged peptide table
mapped = df1_annotated[df1_annotated["Protein ID"].notna()].copy()

# Build a clean Series of UniProt IDs
prot_ids = (
    mapped["Protein ID"]
    .astype(str)                        # make sure it's string
    .str.replace(r"\s+", "", regex=True)  # remove *all* spaces inside
    .str.split(";")                     # split on semicolons
    .explode()                          # turn each element into its own row
)

# Drop empties
prot_ids = prot_ids[prot_ids != ""].dropna()

# Deduplicate
prot_ids = prot_ids.drop_duplicates().reset_index(drop=True)

print("Number of unique IDs:", prot_ids.shape[0])
prot_ids.head(30)

Number of unique IDs: 192


0     P54227
1     O88343
2     P61264
3     P43006
4     P97427
5     P10637
6     O08553
7     Q91XV3
8     Q91V14
9     P28667
10    P10852
11    Q61548
12    Q62261
13    P26645
14    P08553
15    P97546
16    Q5SRX1
17    Q9EPN1
18    Q80YX1
19    P20357
20    Q3UNH4
21    O35274
22    Q6TS35
23    Q8BPN8
24    P51859
25    P14873
26    Q9QYR6
27    P34926
28    P15205
29    Q8R1Q8
Name: Protein ID, dtype: object

In [40]:
# Checks how many entries still contain a semicolon
(prot_ids.str.contains(";")).sum()

np.int64(0)

In [42]:
# Writes file (txt --> must copy and paste into reactome)
prot_ids.to_csv("reactome_protein_list.txt", index=False, header=False)

1. Go to Reactome Pathway Browser
2. Use reactome list of unitprot IDs
3. Run default settings 
4. Download the results

In [51]:
# Bring pathway annotation into Jupyter

reactome = pd.read_csv("reactome_pathway_analysis.csv")
reactome.head(20)

,Pathway identifier,Pathway name,#Entities found,#Entities total,Entities ratio,Entities pValue,Entities FDR,#Reactions found,#Reactions total,Reactions ratio,Species identifier,Species name,Submitted entities found,Mapped entities,Found reaction identifiers
0,R-MMU-9837999,Mitochondrial protein degradation,11,81,0.008756,0.000001,0.000653,8,16,0.001806,10090,Mus musculus,P51881;P97807;Q99KI0;O08749;Q03265;Q9CZU6;Q9D0...,P51881;P97807;Q99KI0;O08749;Q03265;Q9CZU6;Q9D0...,R-MMU-9838035;R-MMU-9838081;R-MMU-9838093;R-MM...
1,R-MMU-9861559,PDH complex synthesizes acetyl-CoA from PYR,4,6,0.000649,0.000009,0.001888,3,3,0.000339,10090,Mus musculus,Q8BKZ9;O08749;Q9D051;Q8BMF4,Q8BKZ9;O08749;Q9D051;Q8BMF4,R-MMU-9861667;R-MMU-9861616;R-MMU-9861734
2,R-MMU-1428517,Aerobic respiration and respiratory electron t...,17,237,0.025619,0.000013,0.001888,31,97,0.010947,10090,Mus musculus,Q8BKZ9;Q9CQQ7;O08749;Q03265;P19783;Q9D051;Q8BM...,Q8BKZ9;Q9CQQ7;O08749;Q03265;P19783;Q9D051;Q8BM...,R-MMU-9864415;R-MMU-9865579;R-MMU-9861734;R-MM...
3,R-MMU-3214847,HATs acetylate histones,7,44,0.004756,0.000047,0.004023,6,8,0.000903,10090,Mus musculus,Q64525;P68433;Q8CGP0;P62806;Q64475;P10853;Q64478,Q64525;P68433;Q8CGP0;P62806;Q64475;P10853;Q64478,R-MMU-3318415;R-MMU-3321883;R-MMU-3321805;R-MM...
4,R-MMU-1592230,Mitochondrial biogenesis,6,30,0.003243,0.000047,0.004023,2,5,0.000564,10090,Mus musculus,P09671;Q9CQQ7;Q03265;P56480;Q91VR2;Q9DB20,P09671;Q9CQQ7;Q03265;P56480;Q91VR2;Q9DB20,R-MMU-5688289;R-MMU-8949580
5,R-MMU-3214815,HDACs deacetylate histones,8,63,0.006810,0.000065,0.004023,5,5,0.000564,10090,Mus musculus,Q64525;P68433;Q8CGP0;P62806;Q64475;P10853;Q6PD...,Q64525;P68433;Q8CGP0;P62806;Q64475;P10853;Q6PD...,R-MMU-9679787;R-MMU-3782637;R-MMU-3769447;R-MM...
6,R-MMU-163210,Formation of ATP by chemiosmotic coupling,5,20,0.002162,0.000073,0.004023,3,3,0.000339,10090,Mus musculus,Q9CQQ7;Q03265;P56480;Q91VR2;Q9DB20,Q9CQQ7;Q03265;P56480;Q91VR2;Q9DB20,R-MMU-164834;R-MMU-164840;R-MMU-164832
7,R-MMU-8949613,Cristae formation,5,20,0.002162,0.000073,0.004023,1,1,0.000113,10090,Mus musculus,Q9CQQ7;Q03265;P56480;Q91VR2;Q9DB20,Q9CQQ7;Q03265;P56480;Q91VR2;Q9DB20,R-MMU-8949580
8,R-MMU-9909649,Regulation of PD-L1(CD274) transcription,6,38,0.004108,0.000170,0.008345,1,1,0.000113,10090,Mus musculus,P68433;Q8CGP0;P62806;Q64475;P10853;Q64478,P68433;Q8CGP0;P62806;Q64475;P10853;Q64478,R-MMU-9909631
9,R-MMU-9841922,MLL4 and MLL3 complexes regulate expression of...,6,42,0.004540,0.000290,0.011448,1,3,0.000339,10090,Mus musculus,P68433;Q8CGP0;P62806;Q64475;P10853;Q64478,P68433;Q8CGP0;P62806;Q64475;P10853;Q64478,R-MMU-9843118


<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    3. Share the data using online repositories
  </h2>
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step A. Use PRIDE linkback</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       1. Show that the dataset already exists in PRIDE and is FAIR-compliant.<br>

<h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:0;">
    <b> Step B. Provide the processed files on GitHub</b>
  </h3> 
   <p style="color: #34495e; font-family: 'Arial', sans-serif; font-size: 15px; margin-top: 18px;">
       1. Upload the cleaned peptide table<br>
       2. Upload the regulated protein list<br>
       3. Upload the pathway/network results<br>

<!-- RESULTS & ANALYSIS SECTION -->
<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    Results & Analysis
  </h2>

  <!-- SECTION 1 -->
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:10px;">
    1. Identify peptides, proteins, and their modifications :
  </h3>

  <p style="color:#34495e; font-family:'Arial',sans-serif; font-size:15px; line-height:1.6; text-indent:25px;">
    Using Dataset 1, all peptide sequences were extracted, cleaned, and deduplicated to remove redundant identifications. After preprocessing, a total of 16,400 unique peptides were obtained. Modification parsing based on embedded PTM tags identified three major PTM classes: phosphorylation (ph), acetylation (ac), and deamidation (de; putative glycosylation). PTM frequency analysis showed that phosphorylation was the predominant modification, followed by deamidation and acetylation. Many peptides contained more than one PTM, indicating coordinated modification events on specific protein regions. Dataset 2 provided additional PTM counts for peptides classified as developmentally regulated; merging the two datasets linked regulated peptides to their corresponding protein IDs and confirmed that a subset of 667 peptides could be confidently mapped to 184 proteins with associated PTM information.
  </p>

  <!-- SECTION 1 -->
  <h3 style="color:#7f8c8d; font-family:'Georgia',serif; font-weight:normal; font-size:22px; margin-top:10px;">
    2. Annotate the data with existing biological knowledge :
  </h3>

  <p style="color:#34495e; font-family:'Arial',sans-serif; font-size:15px; line-height:1.6; text-indent:25px;">
    Protein-level annotation was performed by mapping UniProt IDs to external databases, including UniProtKB, STRING, and Reactome. Functional annotations indicated that a large fraction of the identified proteins participate in cytoskeletal organization, vesicle trafficking, axonal growth, and synaptic signaling, consistent with the neuronal developmental context of the dataset. PTM categories aligned with known biological roles: phosphorylation was enriched among signaling and cytoskeletal regulators; acetylation appeared primarily in chromatin-associated and metabolic proteins; and deamidation was frequently associated with secretory or extracellular proteins. Reactome pathway enrichment further highlighted processes related to axon guidance, microtubule dynamics, neurotransmission, and neuronal differentiation. These annotations collectively support the biological interpretation that coordinated multi-PTM regulation shapes neuronal maturation and structural remodeling during brain development.
  </p>

<!-- DELIVERABLES SECTION -->
<div style="background-color:#f8f9fb; padding:25px; border-radius:12px; box-shadow:0px 4px 8px rgba(0,0,0,0.08);">

  <h2 style="color:#2c3e50; font-family:'Helvetica Neue',sans-serif; font-size:28px; margin-bottom:20px;">
    Lab Deliverables
  </h2>

  <!-- Question 1-->
  <p style="color:#34495e; font-family:'Arial',sans-serif; font-size:15px; line-height:1.6; text-indent:25px;">
    <b>1. Which proteins demonstrated the greatest changes in iTRAQ ratios over the course of 8 days? 21 days? 80 days? Over the entire dataset? </b>
      <br>
Across the developmental time points, the peptides showing the largest iTRAQ ratio changes mapped predominantly to proteins associated with mitochondrial metabolism, chromatin regulation, and cytoskeletal remodeling. Early responses (Day 8) were characterized by strong changes in cytoskeletal and signaling proteins, while by Day 21 several peptides associated with metabolic and mitochondrial regulators displayed substantial shifts. At Day 80, peptides corresponding to enzymes in mitochondrial respiration and chromatin-modifying complexes demonstrated the largest fold changes, reflecting maturation-dependent remodeling of energy metabolism and transcriptional control. When considering the dataset as a whole, proteins contributing peptides with consistently large changes across multiple time points were enriched for mitochondrial degradation pathways, pyruvate dehydrogenase regulation, histone acetylation/deacetylation machinery, and TCA cycle function. These patterns indicate sustained developmental regulation across energy metabolism, epigenetic control, and structural neuronal maturation.

<br><br>

 <!-- Question 2-->
  <p style="color:#34495e; font-family:'Arial',sans-serif; font-size:15px; line-height:1.6; text-indent:25px;">
    <b>2. Are extensive PTM modification (i.e. number of PTMs) and PTM regulation correlated? Do they appear to target the same areas of proteins?</b>
      <br>
There was no strict correlation between the number of PTMs on a peptide and the likelihood of developmental regulation; however, regulated PTMs tended to occur within the same functional regions repeatedly targeted during development. Phosphorylated peptides were overrepresented in proteins belonging to mitochondrial pathways, chromatin-modifying complexes (HATs/HDACs, PRC2, MLL3/4), and metabolic regulators, indicating that specific residues within regulatory domains act as PTM “hotspots.” These hotspots often contained multiple PTMs of the same or different types, suggesting coordinated modification rather than random decoration. In contrast, peptides with many PTMs but no regulation typically mapped to proteins outside major developmental pathways. Overall, PTM density alone did not predict regulation, but regulated PTMs clearly clustered within conserved domains participating in mitochondrial remodeling, chromatin regulation, and metabolic transitions, showing that PTM regulation is region-specific rather than globally distributed.